XMonad/Actions/EasyMotion.hs

```diff
@@ -0,0 +1,332 @@
+      -- Zip window buckets with selection keys, then sort them within screens
+      sortedOverlays :: X [[Overlay]]
+      sortedOverlays = fmap (zipWith (appendChords (maxChordLen c)) (sKeys c) . fmap (sortOn ((wa_x &&& wa_y) . attrs))) winBuckets
+      displayF = displayOverlay f (bgCol c) (borderCol c) (txtCol c) (borderPx c)
+      buildOverlay :: Window -> X Overlay
+      buildOverlay w = do
+        wAttrs <- io $ getWindowAttributes dpy w
+        let r = overlayF c th $ makeRect wAttrs
+        o <- createNewWindow r Nothing "" True
+        return Overlay { rect=r, overlay=o, win=w, attrs=wAttrs, chord=[] }
+  overlays <- fmap concat sortedOverlays
+  status <- io $ grabKeyboard dpy rw True grabModeAsync grabModeAsync currentTime
+  if (status == grabSuccess)
+    then do
+      resultWin <- handleKeyboard dpy displayF (cancelKey c) overlays []
+      io $ ungrabKeyboard dpy currentTime
+      mapM_ (deleteWindow . overlay) overlays
+      io $ sync dpy False
+      releaseXMF f
+      case resultWin of
+        Selected o -> return . Just $ win o
+        _ -> whenJust currentW (windows . W.focusWindow . W.focus) >> return Nothing -- return focus correctly
+    else releaseXMF f >> return Nothing
+
+-- | Display overlay windows and chords for window selection
+selectWindow :: EasyMotionConfig -> X (Maybe Window)
+selectWindow EMConf { sKeys = [] } = return Nothing
+selectWindow conf =
+  -- Make sure there are no duplicates in our key lists
+  case duplicatedKeys of
+    0 -> handleSelectWindow conf { sKeys = map sanitiseKeys (sKeys conf) }
+    _ -> return Nothing
+    where
+      keyList = concat (sKeys conf)
+      keyListLen = length keyList
+      uniqueKeyCount = length . toList . fromList $ keyList
+      duplicatedKeys = uniqueKeyCount - keyListLen
+      -- make sure the key lists don't contain: 'cancelKey, backspace
+      sanitiseKeys = toList . fromList . L.filter (not . flip elem [cancelKey conf, xK_BackSpace])
+
+-- | Take a list of overlays lacking chords, return a list of overlays with key chords
+appendChords :: Int -> [KeySym] -> [Overlay] -> [Overlay]
+appendChords _ [] _ = []
+appendChords maxLen ks overlays =
+  zipWith (\o c -> o { chord=c }) overlays chords
+    where
+      chords = replicateM chordLen ks
+      tempLen = -((-(length overlays)) `div` length ks)
+      chordLen = if maxLen <= 0 then tempLen else min tempLen maxLen
+
+-- | Get a key event, translate it to an event type and keysym
+event :: Display -> IO (EventType, KeySym)
+event d = allocaXEvent $ \e -> do
+  maskEvent d (keyPressMask .|. keyReleaseMask) e
+  KeyEvent {ev_event_type = t, ev_keycode = c} <- getEvent e
+  s <- keycodeToKeysym d c 0
+  return (t, s)
+
+-- | A three-state result for handling user-initiated selection cancellation, successful selection,
+--   or backspace.
+data HandleResult a = Exit | Selected a | Backspace
```

Comment:
This is not exported and you're only using it in conjunction with `Overlay`; is there any reason why `a` is a type variable here?